syntax="proto3";

import "google/protobuf/timestamp.proto";

option go_package = "pos.telemetry";

enum MetricTypes {
  COUNTER = 0;
  GUAGE = 1;
//  HISTOGRAM = 2;
}

service MetricManager {
  //Publisher -> Manager
  rpc MetricPublish(MetricPublishRequest) returns (MetricPublishResponse) {}
  //Collector -> Manager
  rpc MetricCollect(MetricCollectRequest) returns (MetricCollectResponse) {}
}

message MetricPublishRequest
{
  // Max 1,000
  repeated Metric metrics = 1;
}

message MetricPublishResponse
{
  int64 totalReceivedMetrics = 1;
}

message MetricCollectRequest
{
  string requester = 1;
}

message MetricCollectResponse
{
  // Max 100,000
  repeated Metric metrics = 1;
}

message Metric
{
  MetricTypes type = 1;
  google.protobuf.Timestamp time = 2;
  string name = 3;
  oneof value {
    uint64 counterValue = 4;
    int64 guageValue = 5;
//    HistogramData histogramValue = 6;
  }
  repeated Label labels = 6;
}

message Label
{
  string key = 1;
  string value = 2;
}

/*
message HistogramData
{
  message HistogramBucket
  {
    string le = 1; //upper inclusive bound
    uint64 value = 2;
  }
  repeated HistogramBucket buckets = 2;
  int64 sum = 3;
  uint64 count = 4;
}
*/